packages/wdio-allure-reporter/src/reporter.ts

```diff
@@ -398,234 +562,336 @@ export default class AllureReporter extends WDIOReporter {
+        if (!this.currentAllureSpec) {
+            return
         }
-        this.startStep(step.title)
+
+        this._startStep(step.title)
+
         if (step.attachment) {
-            this.addAttachment(step.attachment)
+            this.attachFile(step.attachment.name, step.attachment.content, step.attachment.type || ContentType.TEXT)
         }
-        this.endStep(step.status)
+
+        this._endTest(step.status)
     }
 
     addArgument({
         name,
         value
     }: any) {
-        if (!this.isAnyTestRunning()) {
-            return false
+        if (!this.currentTest) {
+            return
         }
 
-        const test = this._allure.getCurrentTest()
-        test.addParameter('argument', name, value)
+        this.currentTest.addParameter(name, value)
     }
 
-    isAnyTestRunning() {
-        return this._allure.getCurrentSuite() && this._allure.getCurrentTest()
-    }
-
-    isScreenshotCommand(command: CommandArgs) {
-        const isScrenshotEndpoint = /\/session\/[^/]*(\/element\/[^/]*)?\/screenshot/
-        return (
-            // WebDriver protocol
-            (command.endpoint && isScrenshotEndpoint.test(command.endpoint)) ||
-            // DevTools protocol
-            command.command === 'takeScreenshot'
-        )
-    }
+    addAllureStep(metadata: MetadataMessage) {
+        const {
+            attachments = [],
+            labels = [],
+            links = [],
+            parameter = [],
+            steps = [],
+            description,
+            descriptionHtml,
+        } = metadata
 
-    dumpJSON(name: string, json: object) {
-        const content = JSON.stringify(json, null, 2)
-        const isStr = typeof content === 'string'
-        this._allure.addAttachment(name, isStr ? content : `${content}`, isStr ? 'application/json' : 'text/plain')
-    }
+        if (description) {
+            this.addDescription({
+                description,
+                descriptionType: DescriptionType.MARKDOWN
+            })
+        }
 
-    attachScreenshot() {
-        if (this._lastScreenshot && !this._options.disableWebdriverScreenshotsReporting) {
-            this._allure.addAttachment('Screenshot', Buffer.from(this._lastScreenshot, 'base64'))
-            this._lastScreenshot = undefined
+        if (descriptionHtml) {
+            this.addDescription({
+                description: descriptionHtml,
+                descriptionType: DescriptionType.HTML
+            })
         }
+
+        labels.forEach((label) => {
+            this.addLabel(label)
+        })
+        parameter.forEach((param) => {
+            this.addArgument(param)
+        })
+        links.forEach((link) => {
+            this.addLink(link)
+        })
+        attachments.forEach((attachment) => {
+            this.addAttachment(attachment)
+        })
+        steps.forEach((step) => {
+            this.currentAllureSpec!.addStep(AllureCommandStepExecutable.toExecutableItem(this._allure, step))
```

Comment:
We can't skip the nullability check  here because step() can be called from wdio hooks when no test has run yet e.g. `beforeTest` hooks.

In Java adapters, Allure just logs that as an error, like `[main] ERROR io.qameta.allure.AllureLifecycle - Could not start step: no test case running`. Probably we should do the same here.

Reply:
Add the check in the begginging for the method. Destructuring is a hack to bypass typescript notifications